core/packages/contracts/test/Gateway.t.sol

```diff
@@ -0,0 +1,652 @@
+
+        Gateway.UpgradeParams memory params = Gateway.UpgradeParams({
+            impl: address(newLogic),
+            implCodeHash: address(newLogic).codehash,
+            initParams: abi.encode(666)
+        });
+
+        vm.expectRevert(Gateway.SetupFailed.selector);
+        GatewayMock(address(gateway)).upgradePublic(abi.encode(params));
+    }
+
+    function testUpgradeFailCodeHashMismatch() public {
+        GatewayV2 newLogic = new GatewayV2();
+
+        Gateway.UpgradeParams memory params =
+            Gateway.UpgradeParams({impl: address(newLogic), implCodeHash: bytes32(0), initParams: abi.encode(42)});
+
+        vm.expectRevert(Gateway.InvalidCodeHash.selector);
+        GatewayMock(address(gateway)).upgradePublic(abi.encode(params));
+    }
+
+    function testSetOperatingMode() public {
+        Gateway.SetOperatingModeParams memory params =
+            Gateway.SetOperatingModeParams({mode: OperatingMode.RejectingOutboundMessages});
+
+        OperatingMode mode = IGateway(address(gateway)).operatingMode();
+        assertEq(uint256(mode), 0);
+
+        GatewayMock(address(gateway)).setOperatingModePublic(abi.encode(params));
+
+        mode = IGateway(address(gateway)).operatingMode();
+        assertEq(uint256(mode), 1);
+    }
+
+    function testWithdrawAgentFunds() public {
+        deal(assetHubAgent, 50 ether);
+
+        address recipient = makeAddr("recipient");
+
+        bytes memory params = abi.encode(
+            Gateway.TransferNativeFromAgentParams({agentID: assetHubAgentID, recipient: recipient, amount: 3 ether})
+        );
+
+        GatewayMock(address(gateway)).transferNativeFromAgentPublic(params);
+
+        assertEq(assetHubAgent.balance, 47 ether);
+        assertEq(recipient.balance, 3 ether);
+    }
+
+    /**
+     * Assets
+     */
+
+    function testRegisterToken() public {
+        vm.expectEmit(false, false, false, true);
+        emit TokenRegistrationSent(address(token));
+
+        vm.expectEmit(true, false, false, false);
+        emit OutboundMessageAccepted(
+            assetHubParaID, 1, SubstrateTypes.RegisterToken(address(gateway), address(token), bytes2(0x3500))
+        );
+
+        IGateway(address(gateway)).registerToken{value: 2 ether}(address(token));
+    }
+
+    function testSendTokenAddress32() public {
+        // Let gateway lock up to 1 tokens
+        token.approve(address(gateway), 1);
+
+        // Multilocation for recipient
+        ParaID destPara = ParaID.wrap(2043);
+        bytes32 destAddress = keccak256("/Alice");
+
+        vm.expectEmit(true, true, false, true);
+        emit TokenSent(address(this), address(token), destPara, abi.encodePacked(destAddress), 1);
+
+        // Expect the gateway to emit `OutboundMessageAccepted`
+        vm.expectEmit(true, false, false, false);
+        emit OutboundMessageAccepted(
+            assetHubParaID, 1, SubstrateTypes.SendToken(address(gateway), address(token), destPara, destAddress, 1)
```

Comment:
As discussed on Slack, the gateway address will now form part of the asset multilocation.

Reply:
link to Slack convo please? this sounds weird/off to me

Reply:
When executing xcm on bridgehub we `DescendOrigin` to a sovereign account. Before this PR, we created a sovereign for the Registry contract from its address `../../GlobalConsensus(Ethereum{ chain_id }/AccountKey20{ $REGISTRY_CONTRACT_ADDRESS }`.
So this question is me asking if we would use the AssetHubAgent or GatewayProxy contract address for the `DescendOrigin`. I think in earlier conversations we decided on using AssetHubAgent but the tests use GatewayProxy in some places so I was unclear. So we decided to use GatewayProxy as the address for creating a sovereign and I am busy implementing that now.

Reply:
Just adding more background info here:

Tokens bridged over from Ethereum to Polkadot will be stored on the AssetHub (Statemint) parachain. AssetHub considers these assets to be "foreign" and each one is uniquely identified by a MultiLocation.

This MultiLocation must be of the form `/<PREFIX>/<ID>`, where `<PREFIX>` is the consensus system that owns or controls the asset, and `<ID>` is an arbitrary asset id within the consensus system.

AssetHub sets up access controls so that only `<PREFIX>` can mint/burn foreign assets deposited on AssetHub.

What this means is that we need to use a MultiLocation like the following for ERC20s bridged over to AssetHub:
```
/GlobalConsensus(Ethereum)/AccountKey20(Gateway.address)/AccountKey20(Token.address)
```

This could be decomposed into PREFIX and ID
```
PREFIX=/GlobalConsensus(Ethereum)/AccountKey20(Gateway.address)
ID=/AccountKey20(Token.address)
```

Effectively this allows our Gateway contract on Ethereum to administer assets on AssetHub.

Reply:
I'm not sure this is a good idea. This ties the Asset on Polkadot to our particular Gateway. Shouldn't the prefix ideally should just be the asset, ie, 

```
PREFIX=/GlobalConsensus(Ethereum)
ID=/AccountKey20(Token.address)
```

It should be up to the bridge to encapsulate that the Gateway is the only way to act as Ethereum, not baked into the asset?

Reply:
How does AssetHub validate the PREFIX? Is the prefix controlled by Polkadot itself or?

Reply:
Trust asset allowed is defined in AssetHub runtime as
https://github.com/Snowfork/cumulus/blob/9b83c7325df5ec50fec2571f8ff304d40a7e88fe/parachains/runtimes/assets/asset-hub-kusama/src/xcm_config.rs#L399
with trusted reserve locations(i.e. PREFIX) in

Reply:
> This ties the Asset on Polkadot to our particular Gateway

The Gateway is now a proxy so should not change. And I guess it's useful to differ the same asset(e.g. weth) from  other bridge providers(e.g. From Axelar).

Reply:
Thanks @yrong! Yeah we had a quick call and discussed this and basically:
 - We'll have the same issue with XCM Transact and the Identity/Multilocation of agents, since they'll be tied to our particular gateway too. We'll likely need a Gateway prefix for those too
 - On the other hand, there is a disadvantage to fragmenting liquidity and fragmenting identities across different bridges. Ideally the Polkadot community can come to some wider consensus on what their primary trust model is for bridges and pick a bridge or composition of bridges to control the entire Ethereum PREFIX directly, instead of just a particular suffix for a particular contract. If desired, this will likely have to come later via wider Polkadot governance, so it's not something we should implement now but we should be thinking about it and ensuring we'll have future compatibility if this does become needed.